### ISSUE TITLE
Case file for windninja

### ISSUE THREAD

**Issue by ksshannon** [firelab/windninja#7]
Design and implement a case file.
- Case files will be optional.
- Case files will be immutable.
- Users can load an existing file but not save back to it, unless they overwrite the old.
- Output grids will be saved to the file.

**Comment by ksshannon** [comment]
After working on some other things (mostly wx), I really want this.  It's a big change I think though, but simplify so much stuff.  I changed my mind on a few things above, I think we should be allowed to *append* to a case file so it can hold multiple forecasts and outputs.  Once a surface file is written, it can't be changed.  Same with any outputs generated.  One of the issues I forsee is reading directly from the case file, but I think we can figure that out.  We typically don't do too much memory work outside of the solver, so maybe we are okay.  I'm going to write up a schema and an interface, and we can go over it.

**Comment by jforthofer** [comment]
I think a case file would be good also for a lot of reasons.

**Comment by nwagenbrenner** [comment]
We should include some information about the WindNinja version here too. Chuck recently asked for this.

**Comment by RuiZhang-kwf8** [comment]
Finished base version of casefile 
Some improvements - 
*scraping the vtk from ninjafoam
*put domain average run.cfg files in the corresponding time folder with the output vtk for that run.

**Comment by jforthofer** [comment]
Some things to consider here...  Shouldn't we generally follow the naming convention we currently use for output files?  Currently the output file names include information about the initialization method, mesh resolution, date time info, etc...  This basename is already formed properly and would be easy to use.  It also is convenient for users, since then they normally don't have to worry about overwriting files when they do multiple runs, since they would usually be changing something for each run (initialization method, resolution, etc.).  I suggest using the current methodology for naming, but additionally (if we don't already do this) what Natalie mentioned above about asking the user in the GUI if they want to overwrite (only when that would occur) and appending a number for the CLI.

**Comment by nwagenbrenner** [comment]
Yeah, we can do that. The only reason for changing to just the DEM name was to keep the case file names shorter, but maybe that doesn't really matter. We can use the full basename that we use for the other output products. I'll update the list above.

**Comment by jforthofer** [comment]
OK, I just think it's more convenient for users.  That's why we did it like that years ago.  It started when people would do domain average simulations in "batch" type mode in the GUI, we needed to name the output files differently for each run, so we added the input speed and direction to the filenames so they are unique with no user interaction.

**Comment by RuiZhang-kwf8** [comment]
Sorry but I'm a bit confused on the output file name. When you say the "the output file names include information about the initialization method, mesh resolution, date time info, etc..." is that referring to the VTK or a different file generated?

**Comment by nwagenbrenner** [comment]
Basically, you can just use the normal output file name for the case file name. Tutorial 1 has details about the naming convention we use (p. 13):

https://github.com/firelab/windninja/blob/master/doc/tutorials/WindNinja_tutorial1.pdf

For example, for a domain average run, the name contains the DEM name, the input wind direction, the input speed, and the output resolution. There are different protocols depending on the type of initialization, whether it’s a time-varying run, etc. See the set_outputFilenames() function in ninja.cpp and SetOuputFilenames in ninjafoam.cpp for more info. These functions generate the string that is used for all output products (.kmz, .shp, .asc, .pdf, and .vtk) and the one that should be used for the case file basename (we will still use the .ninja extension).

**Comment by jforthofer** [comment]
I was just talking about the normal output files...  for example the Google Earth KMZ files:

If I do a "domain average" initialized run with diurnal turned on (which means a date/time is needed during the simulation for solar angle calcs) the file name is:
missoula_270_10_01-08-2025_0931_471m.kmz  =>  So this name inlcudes the elevation file basename (missoula), input direction (270), input speed (10), date/time (01-08-2025_0931), and the mesh resolution (471m).

Other settings (like initialization method, mesh resolution, diurnal, etc.) can cause changes in this filename form.  For example if I didn't have diurnal turned on in the example above, the date/time would be omitted in the filename.  Again, the reason for all this was that users were doing multiple simulations changes settings, and we wanted to make it convenient for them so they didn't have to come up with unique filenames themselves.

**Comment by RuiZhang-kwf8** [comment]
sounds good

**Comment by nwagenbrenner** [comment]
Remaining items:

- [ ] casefile names still not quite as described above
- [ ] I'm currently getting a segfault in the CLI for domain average runs
- [ ] Need to double check VTK files are always being generated
- [ ] Clean up debug code, ensure proper formatting/coding standards, double check cross-platform functions are being used

**Comment by latwood** [comment]
the vtk file generation needs to be checked again, after a merge with the master branch, the casefile branch is using very outdated ninjafoam vtk output stuff.

I'm currently trying to get the branch cleaned up with the code as it is, then plan to merge the master branch into the casefile branch, update the casefile branch ninjafoam vtk output stuff with the casefile to match the updated master branch code, then merge the casefile branch back into master. But, there's still a LOT of code cleanup left to do. Getting closer now though, trying to do it a little bit at a time.

**Comment by latwood** [comment]
The casefile branch has been thoroughly cleaned up, and is ready for merging processes with the master branch. Then it needs to be tested on Windows.

Because the casefile branch inserted casefile stuff into old vtk methods, amongst other expected merge conflicts, I think it would be best to merge the master branch into the casefile branch, editing the casefile branch to then work again with the changes from the master branch, THEN merge it to the master branch after more testing/edits to get it to work on Windows. Will have discussions with @masonwillman tomorrow about the best way to go about doing the merge process.

**Comment by latwood** [comment]
Successfully merged the cleaned up and working casefile branch into a copy of the master branch called casefile-updated, which is showing it can be merged into the master branch without merge conflicts now. @Mason is reviewing the casefile-updated branch, double checking that it builds and runs on Windows, and then it will be ready to merge with the master branch.

Things that still need done, before the merge:
- [ ] make the writeCaseFile gui option be a regular WindNinjaInputs.cpp option, extending its use to the cli, with the default value to be true
- [ ] make sure everything builds and runs fine on Windows, testing various combinations of each type of run as well as a few tests with combinations of writeCaseFile and writeVtk on or off. The cli should be able to run each of the provided config files, most of these fuller tests should be done in the gui, including, don't forget, a mass solver domainAvg run with two or more wind speeds/directions
- [ ] note, and speaking of not forgetting I almost forgot this one. The momentum solver had troubles with multi-runs in the casefile branch, would be related to reusing a case, if this turns out to still be going on, @latwood will probably need to look into it

Stuff that can be done after the merge, in a new branch copied from the master branch, let's wait till we have a group discussion on these before we try implementing them.
- general double check the formats of names/stuff that goes into the ninja.zip file, I made a few changes to the naming schemes in my cleanup work
- wrap up any, per run, run type input scraping, using the format and style that the cleaned up code now currently has
- scraped inputs from gui WidgetDownloadDEM.cpp and stationFetchWidget.cpp/pointInput.cpp currently just grab the latest/last/most current download info, this might be fine, but do we need to change the style to store like a list of each download set instead? Though not sure how one would keep track of before vs after clicking solve for such lists.
- casefile as well as cli.cpp input processing, need tested and adjusted in whatever ways necessary to allow a casefile to be rerun as a windninja simulation.

**Comment by latwood** [comment]
Ready for testing again.

As far as todos, note the last commit, 00948fd, I ended up not extending the cli and gui writeCaseFile options to be WindNinjaInputs to better handle closing the zip file when WindNinja shuts down unexpectedly.

One other todo, pointInitialization runs currently grab ALL .csv related folders and files seen in the gui tab, even those not related to the run. Not sure how much this is a casefile thing, or just a, can we improve how pointInput.cpp detects which files are station files, type thing.
